### PR TITLE
Switch to official Rollbar client

### LINKED ISSUE #17: Use official go client
https://github.com/rollbar/rollbar-go and couple it with https://github.com/rollbar/rollbar-go/issues/3#issuecomment-353653452

### PR DESCRIPTION
**Pull request by fgblomqvist**
This switches out stvp/roll to rollbar/rollbar-go (closes #17). The main difference is that rollrus now leaves the stack trace extraction/handling to rollbar-go since they (currently) do not provide a method to pass them a stack trace. While the current version of rollbar-go is not able to extract stack traces from pkg/errors, I have submitted a PR that adds this feature: https://github.com/rollbar/rollbar-go/pull/53.

As soon as that PR gets merged (which I hope it does) the end result will be the same as when stvp/roll was used. Probably makes sense to wait with merging this PR until the above PR gets merged.

If anyone want to try out the the new implementation, the master branch of our rollrus fork uses our fork of rollbar-go: https://github.com/monsterroster/rollrus

I have tried it with rollbar.com and the results (in terms of stack trace/message) are great :)

**Please note that this PR also comes with a `go.mod`, and it would make sense to merge #23 first (just for the sake of keeping changes separate).**

### CODE REVIEW

**Comment by fgblomqvist** [comment]
As can be seen in the linked PR, things are a bit on hold in terms of extracting pkg/errors stack traces in rollbar-go. I might have time to add the functionality I proposed in that PR sometime next week/later, but until then we're kinda stuck. I do agree that this should be merged on a v2 branch.

**Comment by freeformz** [comment]
@fgblomqvist Thanks for the follow up. I labeled this `wip` for now. Please ping back when the upstream stuff is handled, esp if you want to talk through v2 options.

**Comment by mback2k** [comment]
Any updates on this issue? To me it looks like the upstream changes to rollbar-go have already been merged.

**Comment by fgblomqvist** [comment]
@mback2k Sorry I just realized I linked the wrong PR, updated the original post. Anyway, as can be read in that PR, it probably won't get merged. Instead, a different solution should be created where it's easy for anyone to configure the error extraction (e.g. through some hook). This is to decrease the number of dependencies. I haven't had the time/priority to do that yet. Maybe I will sometime in the future, unless someone else steps up. Nonetheless, I think the v2 approach is far better than forking with a new name (will just create confusion). So essentially what needs to be done is 1. Submit PR with suggested feature to rollbar-go, and 2. Merge this PR onto a v2 branch. Not too many steps.

**Comment by mback2k** [comment]
@fgblomqvist Thanks for the feedback and update. Could you please explain your latest changes and force-pushes a little bit? Are the changes already usable?

**Comment by fgblomqvist** [comment]
@mback2k Unfortunately the force-push was just for logistics. As in, there have been no new commits (which can be seen by the "5 months ago" timestamp that GitHub shows).

Since this PR seems to create a bit of confusion, it might be better to just close it. I honestly probably won't have the time nor the interest to continue working on this anytime soon due to other obligations. Furthermore, the change that needs to happen is to `rollbar-go` and I would have expected the Rollbar team to step up and maintain that package properly, which has not happened so far. I think it's worth noting that it is used by tons of paying customers, so naturally you would expect the company that profits from it to keep it up-to-date.

**Comment by mback2k** [comment]
@fgblomqvist Thanks a lot for the reply. Before you give up, could you please give an overview what needs to be done. I may be able to perform some of the required work for a PR.

File: rollrus.go [hunk 1]
```diff
@@ -145,22 +145,20 @@ func setupLogging(token, env string, levels []logrus.Level) {
...
-// ReportPanic attempts to report the panic to rollbar using the provided
+// ReportPanic attempts to report the panic to Rollbar using the provided
 // client and then re-panic. If it can't report the panic it will print an
 // error to stderr.
 func (r *Hook) ReportPanic() {
 	if p := recover(); p != nil {
-		if _, err := r.Client.Critical(fmt.Errorf("panic: %q", p), nil); err != nil {
-			fmt.Fprintf(os.Stderr, "reporting_panic=false err=%q\n", err)
-		}
+		r.Client.ErrorWithLevel(rollbar.CRIT, fmt.Errorf("panic: %q", p))
```
Comment (mback2k) [inline rollrus.go]: I think there should be r.Client.Wait before the final panic() to make sure that the error actually reaches Rollbar.

File: rollrus.go [hunk 2]
```diff
@@ -192,50 +192,51 @@ func (r *Hook) Fire(entry *logrus.Entry) error {
...
-	case hasTrace && level == logrus.PanicLevel:
-		_, err = r.Client.CriticalStack(cause, trace, m)
-	case hasTrace && level == logrus.ErrorLevel:
-		_, err = r.Client.ErrorStack(cause, trace, m)
-	case hasTrace && level == logrus.WarnLevel:
-		_, err = r.Client.WarningStack(cause, trace, m)
 	case level == logrus.FatalLevel || level == logrus.PanicLevel:
-		_, err = r.Client.Critical(cause, m)
+		skip := framesToSkip(2)
+		r.Client.ErrorWithStackSkipWithExtras(rollbar.CRIT, cause, skip, m)
```
Comment (mback2k) [inline rollrus.go]: Same here, since this quits the program, I think there should be r.Client.Wait to make sure that the error actually reaches Rollbar.

**Comment by mback2k** [comment]
@fgblomqvist I just posted a [sub-PR](https://github.com/fgblomqvist/rollrus/pull/1) with updates to this PR.

**Comment by freeformz** [comment]
Actually, nevermind, I see what you did that: rollbar-go takes map[string]interface{} in a bunch of places. I think that is a mistake, but oh well.

**Comment by mback2k** [comment]
Hm, I am a little sad that so many restructuring was done in this repo before this PR landed, now everything needs to get rebased again...

**Comment by fgblomqvist** [comment]
Yeah it's unfortunate. I'll have time to rebase it sometime next week, but not before then.

**Comment by fgblomqvist** [comment]
It's probably quite obvious that rebasing this PR onto the current master was a shitshow. There's still a test failing that I just can't bother fixing, but other than that it might be okay. I can't say for sure since it felt like one big hack reapplying my changes.

@freeformz 
To take a previously poorly managed repo, with a major PR pending, and just suddenly blatantly ignore everything and just toss all the code around without any real benefit, is just extremely poor management skills. What you did is only for looks/code structure and could have waited until this PR was merged. At the very least, you should have notified this PR thread about your planned changes. The fact that this happens to a repo that is backed by a major company (that I assume you work for) is shocking, can't be fun working on the rollbar team if that is how their codebase is managed. In my mind, you don't just sacrifice volunteers' time like this, but I guess that doesn't matter when only $$$ does.

I have no more interest in helping out with this. You guys are welcome to take these changes or throw them away, I can't care anymore. Companies shouldn't dabble with open source if they don't know what they're doing.

**Comment by freeformz** [comment]
@fgblomqvist You are correct and I messed up. I'm sorry. :-( How can I make it up to you? 

The failing test is because the old extractError function used to extract the cause, but the new one doesn't. I don't think it makes sense for the current extractError to extract the cause (which is how it is), so this test should probably be changed.

I think the test should be something like this:

```golang
func TestExtractErrorWithCause(t *testing.T) {
	entry := logrus.NewEntry(nil)
	entry.Data["err"] = errors.Wrap(io.EOF,"foo bar baz")

	err := extractError(entry)
	expected := "foo bar baz: EOF"
	if got := err.Error(); got != expected {
		t.Fatalf("got %q, wanted %q", got, expected)
	}
}
```

If you want to replace the test I will accept this. I tagged the current master as v0.0.1, so Go modules users (and probably others) should be fine for now.

**Comment by freeformz** [comment]
I opened that up as a PR here: https://github.com/fgblomqvist/rollrus/pull/2

**Comment by fgblomqvist** [comment]
@freeformz Apology accepted. I merged your PR, thanks. I merged latest master into this PR as well, which did bring with it another test for cause extraction. Don't think that is needed with this branch so deleted it. Now all tests pass (about 80% coverage, but all vital parts seem to be covered, but might still be some case out there that should be tested).

Feels like it would be good to test this in a real application as well (since I haven't done that since the rebase), but if you think the code looks good enough and makes sense then I suppose this can be merged.

**Comment by mback2k** [comment]
@fgblomqvist Are you sure that you covered all the changes that I made during my [PR](https://github.com/fgblomqvist/rollbar-go/pull/1/files) in your rebase? Your commit https://github.com/heroku/rollrus/pull/24/commits/9097165cfecd92e94919d257ad053bee5e94642b looks a lot smaller than mine https://github.com/fgblomqvist/rollbar-go/pull/1/commits/a9a01a2f53af11dd0c59e7007f90f274c00632e3 and https://github.com/fgblomqvist/rollbar-go/pull/1/commits/58fd58663ce2f38a98e549552588a7ebbb91c5b5.

**Comment by fgblomqvist** [comment]
@mback2k I wouldn't be surprised if something fell off due to the mess. As you can see, I had to manually add your changes back afterwards since (if I remember it correctly) they somehow got lost during the rebase. I don't know how I would add you to the author line of a relevant commit, but I'll look into it and take a look at if something disappeared. I don't have time today but hopefully one of these coming days. Again, if the rebase wouldn't have been a mess I would, of course, have just kept your commit in there. Another option is that I just toss away 9097165cfecd92e94919d257ad053bee5e94642b and let you rebase your PR on this branch, and then I can just merge a PR from you again. That sounds like probably the easiest option?

**Comment by mback2k** [comment]
@fgblomqvist Sorry, I screwed up with my previous comment. I somehow compared the current PR with the changes to rollbar-go and not rollrus. Let me take another look at this current PR and test with with my own software. I will then report back.

And regarding the author line, just for the future, you could take a look at the [--author option for git commit](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---authorltauthorgt) (--amend).

**Comment by fgblomqvist** [comment]
@mback2k Fixed the history (lmk if this is alright). Now your authorship is preserved (and I also clarified in the commit message that I just rebased your changes) :)

**Comment by fgblomqvist** [comment]
Alright I think this is done now. Author history is correct, it's up to date with master, and all tests are passing. More testcases are of course welcome. But other than that, this branch should stay clean from now on (no more fixing/rebasing should be needed).

**Comment by mback2k** [comment]
Due to the ongoing rebase, my other comment on https://github.com/fgblomqvist/rollrus/commit/9097165cfecd92e94919d257ad053bee5e94642b was lost here, so I am re-adding that as a quote:

> @fgblomqvist Thanks for rebasing my commit mback2k@4fb626a. 😄
> 
> So I went ahead and verified the PR rebase manually by looking at each individual line change. To me the rebase looks good and nothing seems to have been lost during rebase, at least to me. Also all tests pass for me and I was able to use the updated PR branches with a local software.

Thanks again for the rebase and fixing the authorship, both looks good to me. I think this PR is ready to go. 👍

**Comment by freeformz** [comment]
Thank You!

**Comment by freeformz** [comment]
I tagged this, with a small fix for gofmt stuff, as v0.1.0

**Pull request merged**